sale_loyalty_general_discount_promo_code/models/sale_order.py

```diff
@@ -0,0 +1,112 @@
+import random
+from odoo import models, _
+from collections import defaultdict
+from odoo.tools import format_amount
+
+def _generate_random_reward_code():
+    return str(random.getrandbits(32))
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def get_total_saving_specific(self, reward):
+        self.ensure_one()
+        assert reward.discount_applicability == 'specific'
+        total_saving = 0
+        lines_to_discount = self.env['sale.order.line']
+        order_lines = self.order_line - self._get_no_effect_on_threshold_lines()
+        for line in order_lines:
+            if not line.product_uom_qty or not line.price_total:
+                continue
+            domain = reward._get_discount_product_domain()
+            if not line.reward_id and line.product_id.filtered_domain(domain):
+                lines_to_discount |= line
+
+        for line in lines_to_discount:
+            line.discount = reward.discount
+            if line.reward_id:
+                continue
+            pricelist_price = line._get_pricelist_price()
+            total_saving += pricelist_price * (reward.discount / 100)
+        return total_saving
+
+    def get_total_saving_cheapest(self, reward):
+        self.ensure_one()
+        assert reward.discount_applicability == 'cheapest'
+
+        cheapest_line = self._cheapest_line()
+        if not cheapest_line:
+            return 0
+        pricelist_price = cheapest_line._get_pricelist_price()
+        return pricelist_price * (reward.discount / 100)
+
+    def get_total_saving_order(self, reward):
+        self.ensure_one()
+        assert reward.discount_applicability == 'order'
+        total_saving = 0
+        for line in self.order_line:
+            if line.reward_id:
+                continue
+            pricelist_price = line._get_pricelist_price()
+            total_saving += pricelist_price * (reward.discount / 100)
+        return total_saving
+
+    def get_total_saving(self, reward):
+        total_saving = 0
+        reward_applies_on = reward.discount_applicability
+        if reward_applies_on == 'order':
+            total_saving = self.get_total_saving_order(reward)
+        elif reward_applies_on == 'specific':
+            total_saving = self.get_total_saving_specific(reward)
+        elif reward_applies_on == 'cheapest':
+            total_saving = self.get_total_saving_cheapest(reward)
+        return total_saving
+    
+    def _get_reward_values_discount(self, reward, coupon, **kwargs):
+        rewards = super()._get_reward_values_discount(reward, coupon, **kwargs) 
+        currency = self.pricelist_id.currency_id
+
+        for reward_line in rewards:
+            if (
+                reward.reward_type == 'discount'
+                and reward.discount_mode == 'percent'
+                and reward.program_id.program_type == 'coupons'
+            ):
+                saved_amount = self.get_total_saving(reward)
```

Comment:
I don't check it. But maybe same here
```suggestion
                saved_amount = float_around(self.get_total_saving(reward), rounding=currency.rounding
                
```

Reply:
`currency.rounding` has already used in `format_amount`